File: typescript/src/stacks_queues/validParentheses.ts
const PAIRS: Record<string, string> = { ")": "(", "]": "[", "}": "{" };

export function validParentheses(s: string): boolean {
  const stack: string[] = [];
  for (const c of s) {
    if ("([{".includes(c)) {
      stack.push(c);
    } else if (c in PAIRS) {
      if (!stack.length || stack.pop() !== PAIRS[c]) return false;
    } else {
      return false;
    }
  }
  return stack.length === 0;
}

File: python/src/stacks_queues/valid_parentheses.py
"""Valid parentheses for ()[]{} using a stack."""


def valid_parentheses(s: str) -> bool:
    pairs = {")": "(", "]": "[", "}": "{"}
    stack: list[str] = []
    for c in s:
        if c in "([{":
            stack.append(c)
        elif c in pairs:
            if not stack or stack.pop() != pairs[c]:
                return False
        else:
            return False
    return len(stack) == 0
